docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+    - So best-effort pods will be killed if they conflict with burstable pods using less than requested memory
+    - If a burstable pod uses more memory than requested, its OOM_SCORE > 995
+    - If a burstable pod that uses more memory than requested conflicts with a burstable pod using less memory than requested, the former will be killed
+    - If a burstable pod that uses more memory than requested conflicts with a best-effort pod, the behavior is undefined
+- Pod infrastructure container
+  - OOM_SCORE_ADJ: -999
+- Kubelet, Docker, Kube-Proxy
+  - OOM_SCORE_ADJ: -1000 (won’t be OOM killed)
+  - Hack, in the future, we should place all pods into a separate cgroup, and set a limit on the memory they can consume
+
+Ensuring resource guarantees
+- After being configured suitably, the kernel will handle resource guarantees as specified in QoS Functionality
+
+## Issues and Assumptions
+
+- The QoS proposal assumes that swap memory is disabled. If swap is enabled, then resource guarantees (for pods that specify resource requirements) will not hold. For example, suppose 2 top-tier pods have reached their memory limit. They can start allocating memory on swap space. Eventually, if there isn’t enough swap space, processes in the pods might get killed. TODO: ensure that swap space is disabled on our cluster setups scripts.
+- We cannot set OOM_SCORE_ADJ of a process until it has launched. This could lead to races. For example, suppose that a burstable pod is using 70% of the system’s memory, and another burstable pod is using 30% of the system’s memory. A best-effort pod attempts to launch on the Kubelet. Initially the best-effort pod is using 2% of memory, and has an OOM_SCORE_ADJ of 0. So its OOM_SCORE is lower than the burstable  pod using 70% of system memory. The burstable pod will be evicted by the best-effort pod. Short-term TODO: Consider implementing a restart policy where best-effort pods are immediately evicted if OOM killed, but burstable pods are given a few retries. Long-term TODO: push support for OOM scores in cgroups to the upstream Linux kernel. This is out of scope.
+- We need to deal with the Kubelet’s restart policy when a process gets killed by the kernel. Suppose that we have 30 best-effort pods wanting to use 1GB of memory, with an “always” restart policy, in a system with 8GB of memory. In the current proposal, the system will keep launching best-effort pods, and the kernel will keep killing best-effort pods. In fact, the OOM killer is biased towards killing pods that are consuming more memory, so best-effort pods that are running would be evicted. In the first iteration, we will encourage users not to use the “always” restart policy with best-effort pods, but won’t handle the restart issue in Kubelet. Extension: In the final implementation, we will not allow best-effort pods to restart, and will limit the number of times burstable pods can restart if OOM killed.

```

Comment:
Actually, transitioning the pod to Failed phase is an interesting question. @vmarmol and @vishh had good examples of pods where some containers might be more important than others. For example a logging container could run as "best-effort". We might not want to kill the whole pod because a single best-effort container died.

There are a few different schemes possible. We might restart the entire pod, or transition the entire pod into failed state if a container fails. Or we might choose to do this only if a (memory/cpu) guaranteed container fails (if a best-effort container failed, then it probably wasn't very critical). Adding in some discussion of this into the proposal, but I probably won't implement it in the first iteration of the code.